src/libopenrave/trajectory.cpp

```diff
@@ -21,38 +21,175 @@
-            it->second->Serialize(newwriter,options);
-        }
-        writer->Serialize(O);
+    else
+    {
+        s.clear();
     }
-    O << "</trajectory>" << endl;
+    return !!f;
+}
+
+inline bool ReadBinaryVector(std::istream& f, std::vector<dReal>& v)
+{
+    // Get number of data points
+    uint32_t numDataPoints = 0;
+    ReadBinaryUInt32(f, numDataPoints);
+    v.resize(numDataPoints);
+
+    // Load binary directly to vector
+    const uint64_t vectorLengthBytes = numDataPoints*sizeof(dReal);
+    f.read((char*) &v[0], vectorLengthBytes);
+
+    return !!f;
+}
+
+// New feature: Store trajectory file in binary
+void TrajectoryBase::serialize(std::ostream& O, int options) const
+{
+    // NOTE: Ignore 'options' argument for now
+
+    // Write binary file header
+    WriteBinaryUInt16(O, MAGIC_NUMBER);
+    WriteBinaryUInt16(O, VERSION_NUMBER);
+
+    /* Store meta-data */
+
+    // Indicate size of meta data
+    const ConfigurationSpecification& spec = this->GetConfigurationSpecification();
+    const uint16_t numGroups = spec._vgroups.size();
+    WriteBinaryUInt16(O, numGroups);
+
+    FOREACHC(itgroup, spec._vgroups)
+    {
+        WriteBinaryString(O, itgroup->name);   // Writes group name
+        WriteBinaryInt(O, itgroup->offset);    // Writes offset
+        WriteBinaryInt(O, itgroup->dof);       // Writes dof
+        WriteBinaryString(O, itgroup->interpolation);  // Writes interpolation
+    }
+
+    /* Store data waypoints */
+    std::vector<dReal> trajectoryData;
+    this->GetWaypoints(0, this->GetNumWaypoints(), trajectoryData);
+    WriteBinaryVector(O, trajectoryData);
 }
 
 InterfaceBasePtr TrajectoryBase::deserialize(std::istream& I)
 {
+    // Check whether binary or XML file
+    stringstream::streampos pos = I.tellg();  // Save old position
+    uint16_t binaryFileHeader = 0;
+    ReadBinaryUInt16(I, binaryFileHeader);
+
+    // Read binary trajectory files
+    if (binaryFileHeader == MAGIC_NUMBER)
+    {   
+        uint16_t versionNumber = 0;
+        ReadBinaryUInt16(I, versionNumber);
+
+        if (versionNumber != VERSION_NUMBER)
+        {
+            throw OPENRAVE_EXCEPTION_FORMAT(_("unsupported trajectory format version %d "),versionNumber,ORE_InvalidArguments);
+        }
+            
+        /* Read metadata */
+
+        // Read number of groups
+        uint16_t numGroups = 0;
+        ReadBinaryUInt16(I, numGroups);
+
+        ConfigurationSpecification spec;
+        spec._vgroups.resize(numGroups);
+        FOREACH(itgroup, spec._vgroups)
+        {
+            ReadBinaryString(I, itgroup->name);             // Read group name
+            ReadBinaryInt(I, itgroup->offset);              // Read offset
+            ReadBinaryInt(I, itgroup->dof);                 // Read dof
+            ReadBinaryString(I, itgroup->interpolation);    // Read interpolation
+        }
+        this->Init(spec);
+
+        /* Read trajectory data */
+        std::vector<dReal> trajectoryData;
```

Comment:
@ShaysRebellion please do this without creating a temporary `vector`

Reply:
@rdiankov please see above discussion